### PR TITLE
Support zero-byte reads on HTTP response streams

### LINKED ISSUE #61475: SocketsHttpHandler: Support zero-byte read on response streams
Supporting zero-byte read helps users of these APIs reduce memory usage. It allows them to release their read buffers when no data is currently available.

We should support this in our HTTP response streams in SocketsHttpHandler, including RawConnectionStream since this is used for WebSockets and tunneling scenarios.

Ideally we would also release our own read buffer (if any and if possible) and issue a zero-byte read against the underlying connection. This is what SslStream does (as of 6.0). The idea here is that the user has indicated their interest in minimizing buffer memory usage, and so we should do the same for our own internal buffers.

I split this issue off from https://github.com/dotnet/runtime/issues/61223.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #61475

This PR does the plumbing to get zero-byte reads working but doesn't change the buffer management of underlying connections.
On HTTP/1.1: The 4k receive buffer remains
On HTTP/2: We don't issue zero-byte reads to the underlying connection at all, but that is unrelated to whether we support zero-byte reads on the response streams

The minimal change to get WebSockets working on HTTP/1.1 is just 25029a768caa064eb3dd8ebc31dd030a38bb2813

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs [hunk 1]
```diff
@@ -235,6 +288,15 @@ private async Task CopyToAsyncCore(Stream destination, CancellationToken cancell
                 }
             }
 
+            private bool PeekChunkFromConnectionBuffer()
+            {
+                ReadOnlyMemory<byte> bytesRead = ReadChunkFromConnectionBuffer(maxBytesToRead: 0, cancellationRegistration: default);
+
+                // ReadChunkFromConnectionBuffer will return 'default' (NullRef with Length=0) if no data is available or
+                // _connection.RemainingBuffer.Slice(0, bytesToConsume) if there is.
+                return !Unsafe.IsNullRef(ref MemoryMarshal.GetReference(bytesRead.Span));
```
Comment (MihaZupan) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:297]: I wanted to avoid changing the parsing loop below too much here. Should we be using something like a sentinel buffer object to signal available data instead?

**Comment by CarnaViire** [comment]
HTTP/3 part LGTM. I notice `MockQuicStreamConformanceTests.ZeroByteRead_BlocksUntilDataAvailableOrNops` test started failing -- is that after changes to make your added HTTP/3 test work? We don't really support Mock Quic now and will most likely remove it in the near future, so don't spend too much time investigating that...

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs [hunk 2]
```diff
@@ -81,15 +94,30 @@ public override int Read(Span<byte> buffer)
                         return bytesRead;
                     }
 
+                    if (buffer.Length == 0)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:97]: Why is this necessary? Won't the Fill call immediately below take care of this?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs [hunk 3]
```diff
@@ -167,15 +205,30 @@ private async ValueTask<int> ReadAsyncCore(Memory<byte> buffer, CancellationToke
                             return bytesRead;
                         }
 
+                        if (buffer.Length == 0)
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:208]: Same comment as sync above

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 1 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:297]: My suggestion would be to use a bool to explicitly indicate EOF and return this from ReadChunkFromConnectionBuffer. That matches what we do in other cases like HTTP2 (TryReadFromBuffer) etc.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 4]
```diff
@@ -1747,7 +1745,7 @@ private ValueTask<int> ReadBufferedAsync(Memory<byte> destination)
             // If the caller provided buffer, and thus the amount of data desired to be read,
             // is larger than the internal buffer, there's no point going through the internal
             // buffer, so just do an unbuffered read.
-            return destination.Length >= _readBuffer.Length ?
+            return destination.Length == 0 || destination.Length >= _readBuffer.Length ?
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1748]: Why does the 0 byte read case here go through the non-buffered path?

At the very least we should update the comment to explain this.

File: src/libraries/System.Net.Http/tests/FunctionalTests/ResponseStreamZeroByteReadTests.cs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+using System.Net.Test.Common;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Http.Functional.Tests
```
Comment (geoffkizer) [inline src/libraries/System.Net.Http/tests/FunctionalTests/ResponseStreamZeroByteReadTests.cs:20]: Is there any way we could just use stream conformance tests directly here?

We have some support for running stream conformance tests against response streams already. Can we just improve that so it (a) enables zero byte read tests and (b) covers all the cases we want it to?

**Comment by geoffkizer** [comment]
Generally LGTM, a couple comments above.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:97]: Fill will pass the entire connection buffer to the transport stream.
Checking for 0 here here means that we will attempt a zero byte read first if the user requested it and no data was available in the buffer.

This allows us to take advantage of the zero byte read on SslStream.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 4 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:1748]: Same as with the chunked stream, it's to ensure the zero byte read is propagated to SslStream.
BufferedRead will always request at least 4k bytes.

I will update the comment.

**Comment by MihaZupan** [comment]
@CarnaViire the test hang was my doing - I didn't update the conformance test to indicate the stream now supports zero-byte reads.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: > I assume we do this for the non-chunked cases too?

Yes, we now do this for all HTTP 1 response streams.
On HTTP/2 and HTTP/3, the response streams also block on zero-byte reads, but those reads are separate from the reads to the underlying transport.

I agree that the most obvious benefits are to idle-ish connections (streaming), but even regular content transfers will often be slow enough for the receive loop to outperform and be forced to wait.

AspNetCore separates this into two parts:
- The streams (pipes) exposed to the user will block on zero-byte reads
- Whether the underlying transport uses zero-byte reads is controlled by the `WaitForDataBeforeAllocatingBuffer` flag which always defaults to `true`. Their [`SocketTransport`](https://github.com/dotnet/aspnetcore/blob/465f9294d0b9f5403ab1a25c42ca9cf75e3cdeae/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L144-L153) will issue a zero-byte read before every actual read.
  - We should consider proactively doing the same for HTTP/2 and 3 (it would help with #61223).

The overhead of honoring the user's request to issue a zero-byte read should be low, especially if we already checked that no data has been buffered.

File: src/libraries/System.Net.Http/tests/FunctionalTests/ResponseStreamZeroByteReadTests.cs (hunk 5 above)
Comment (MihaZupan) [reply src/libraries/System.Net.Http/tests/FunctionalTests/ResponseStreamZeroByteReadTests.cs:20]: Technically, yes, but it would require changes in the base conformance test class and our loopback servers.

The main issues I hit here were:
- The connected stream tests don't advertise the amount of data that will be transferred to different implementations. For `Content-Length`, we have to know the number upfront.
- The generic loopback server does not support partial response content read/writes, without which we can't have a stream wrapping a `GenericLoopbackConnection`.

We can certainly address all of such problems and turn all of these into `StreamConformanceTests`, I just wasn't convinced it was worth doing as part of this PR.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: > even regular content transfers will often be slow enough for the receive loop to outperform and be forced to wait.

Yes, that's a good point, and I agree it's reasonable to do a zero-byte read in these cases.

To step back for a second, here's how I think about this generally:

The response streams should always support zero-byte reads -- meaning, a zero-byte read will block until data is actually available and the subsequent read is guaranteed to not block. This enables a user to reduce their overall buffer memory usage: they only need to allocate a buffer when data is actually available.

That said, supporting zero-byte read does not require that the response stream does a zero-byte read against the underlying stream. This is most obvious in the HTTP2 case because of multiplexing. But it also applies to a case like the chunked response stream: we (usually) have an internal read buffer, and so we could just do regular (non-zero-byte) reads into that read buffer and then use that to provide the zero-byte-read semantics to the user.

What that approach doesn't address is, being smart about our own buffer management. Currently, we never release the internal read buffer between read calls, even when it's empty. And currently we never do a zero-byte read against the underlying stream. But if the user issues a zero-byte-read against us, then it seems reasonable to assume that they want to trade off some small CPU overhead for reduced memory usage -- that's the whole point of doing a zero-byte read. And so when that happens, we should avoid holding/allocating a buffer during the user's zero-byte read when possible, and instead do a zero-byte read against the underlying stream, and only allocate a buffer when that zero-byte read returns. (That's the change we made to SslStream in 6.0.)

What's weird about this PR is that we haven't changed the internal buffer management at all -- we still never release the internal read buffer. So even though we do a zero-byte read against the underlying stream, we aren't actually getting any advantages from this in terms of reduced memory usage.

Put differently: there's no point in doing a zero-byte read unless we are going to defer allocation of the read buffer until that zero-byte read completes.

Thoughts?

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: > there's no point in doing a zero-byte read unless we are going to defer allocation of the read buffer until that zero-byte read completes

On Windows we'd end up avoiding pinning, yes?

Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: > On Windows we'd end up avoiding pinning, yes?

Yes, that's true. I was hoping you wouldn't notice that :)

That said, I'm not sure how much that matters in practice, particularly if we assume that we *do* implement smart buffer management -- meaning, do zero-byte reads against the underlying stream and don't hold buffers when no data is available.

The interesting case to consider here is something like this: we have read a partial TLS frame into our read buffer. We need to read more data in order to be able to decrypt the entire frame; until we read that, we can't complete the user's zero byte read because we don't have any available data yet.

Should we perform a zero-byte read against the underlying stream in this case? Doing so would avoid pinning on Windows, which is good. However, since we already have a partial frame, it's almost certain that the rest of the frame has already arrived in the kernel buffer, or at worst, will arrive very soon. Sure, in weird cases we could have packet loss or just a very slow pipe and we'd end up pinning the buffer for a non-trivial amount of time. But that seems rare, and it's not at all clear that the benefit of avoiding pinning in this case is worth the cost of the zero-byte read.

And of course there's no pinning on Linux so in that case the zero-byte read is pointless.

(I used SslStream as the example because it's most relevant there. For chunked response streams, the equivalent would be that we have a partial chunk header -- we only need the chunk header plus one byte of the chunk to have data available. But chunk headers are pretty small in general so the impact on the chunked response stream is small.)

If we care about avoiding pinning of buffers on Windows, then there are a lot of cases that cause pinning beyond just the ones described above, and we should look at ways to address the problem holistically, like having pools of perma-pinned or native buffers.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: > we aren't actually getting any advantages from this in terms of reduced memory usage

We are giving the user the option to defer allocating their buffers and avoid allocating the 32k SslStream buffer.
In reality, the user buffer is often also orders of magnitudes larger than the 4k one we control.

I agree we should be smarter about our buffer management, particularly in these zero-byte read cases.
But the benefits enabled by this PR do not hinge on that happening - the user is getting 90% with just this change.

> Put differently: there's no point in doing a zero-byte read unless we are going to defer allocation of the read buffer until that zero-byte read completes.

I view these improvements as two separate changes:
1. Block on zero-byte reads to the response content, translating them into zero-byte reads to the underlying transport where it makes sense
2. Improve our own buffer management - e.g. returning the 4k buffer we own in HttpConnection

If you believe it's a better idea, I can include changes for 2. in this PR as well, but to me, it seemed easier to reason about the changes as two separate PRs.

---

> if the user issues a zero-byte-read against us, then it seems reasonable to assume that they want to trade off some small CPU overhead for reduced memory usage -- that's the whole point of doing a zero-byte read. And so when that happens, we should avoid holding/allocating a buffer during the user's zero-byte read when possible, and instead do a zero-byte read against the underlying stream, and only allocate a buffer when that zero-byte read returns. (That's the change we made to SslStream in 6.0.)

Hard agree, we can be smarter here.

In a lot of cases, we don't need to hold on to the 4k buffer at all:
- For `Raw` (e.g. WebSockets) or `ContentLength` streams, we can (and already do) make use of the user-provided buffer for reads to the transport - we never need our own buffer
- For `Chunked`, we can release the buffer on a zero-byte read. We also [already have specialized logic](https://github.com/dotnet/runtime/blob/a4cab7d3206027566dc14672669caeac56ef8400/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs#L148-L168) where we will use the user-provided buffer if we have nothing buffered and know we are expecting chunk data. We shouldn't be holding onto our buffer in this case. 
- For HTTP/2, I already mentioned above that I believe we should be doing proactive zero-byte reads to the transport (akin to AspNetCore's `WaitForDataBeforeAllocatingBuffer` option). Unless we see there is a measurable performance difference in doing so, I don't think it's even worth exposing it as an option.

I'm happy to make such changes, but I don't believe they necessarily overlap with this PR too much.

Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: I'm fine having two different PRs as you describe.

To get specific here: why are we adding this particular piece of logic here -- lines 95 through 103 -- as part of this PR instead of the follow-on PR?

(a) It's not necessary to make the user's zero-byte read work
(b) It seems intended to support the follow-on PR work to improve buffer management, but it doesn't actually change any buffer management
(c) The comment is confusing because it talks about "defer buffer allocation" but doesn't actually do that

So why make this change in this PR?

Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: Re your comments on buffer improvements to Raw/Chunked/HTTP2/etc, that all seems good to me.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: It's so that the user's zero-byte read translates into a zero-byte read to SslStream (deferring its 32k buffer allocation) while avoiding changes to the lifetimes of `HttpConnection` owned buffers. I.e. it's a minimal change to get most of the benefit, matching other HTTP/1.x streams.
It's akin to [this change](https://github.com/dotnet/runtime/pull/61913/files#diff-595f700a4c83b735d856fe14b5ebe8c7f7c18ae6e586adc8b398a997a0678274R1732) (HttpConnection line 1732) - not needed for the zero-byte read to block, but allows us to make use of SslStream's zero-byte read improvements.

I agree the comment is misleading now without further changes, I'll update that.

Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: Okay, makes sense. Let's change the comment and get this in.

**Review by geoffkizer — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: Just to ensure we are on the same page:

It seems like there are really three pieces of work here to fully support zero-byte reads:

(1) Support user zero-byte read so they can be smart about managing their buffers
(2) Use the user's zero-byte read to improve our own internal buffer management
(3) Perform zero-byte read against the underlying stream so that the underlying stream can be smart about their buffer management

This PR contains (1) and (3) for all response streams (except (3) for HTTP2, due to multiplexing).

We will look at (2) in a future PR. This may include changes to buffer management beyond just zero-byte-read cases. 

We will also look at (3) for HTTP2 in a future PR.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs:95]: Yes, that matches my understanding exactly
